src/packages/emmett-mongodb/src/eventStore/mongoDBEventStore.ts

```diff
@@ -0,0 +1,272 @@
+    if (!stream) throw new Error('Failed to create stream');
+
+    assertExpectedVersionMatchesCurrent(
+      stream.events.length,
+      options?.expectedStreamVersion,
+      MongoDBEventStoreDefaultStreamVersion,
+    );
+
+    // @ts-expect-error The actual `EventType` is different across each stream document,
+    // but the collection was instantiated as being `EventStream<Event>`. Unlike `findOne`,
+    // `findOneAndUpdate` does not allow a generic to override what the return type is.
+    const updatedStream: WithId<EventStream<EventType>> | null =
+      await this.collection.findOneAndUpdate(
+        {
+          streamName: { $eq: streamName },
+          events: { $size: stream.events.length },
+          updatedAt: new Date(),
+        },
+        { $push: { events: { $each: eventCreateInputs } } },
+        { returnDocument: 'after' },
+      );
+
+    if (!updatedStream) {
+      const currentStream = await this.collection.findOne({
+        streamName: { $eq: streamName },
+      });
+      throw new ExpectedVersionConflictError(
+        currentStream?.events.length ?? -1,
+        stream.events.length,
+      );
+    }
+
+    const { streamType, entityId } = fromStreamName(streamName);
+
+    await executeProjections(
+      {
+        streamName,
+        streamType,
+        entityId,
+        streamVersion: updatedStream.events.length,
+        events: updatedStream.events,
+      },
+      options?.projections,
+    );
+
+    return {
+      nextExpectedStreamVersion: updatedStream.events.length,
+      createdNewStream,
+    };
+  }
+}
+
+export const getMongoDBEventStore = async (
+  options: MongoDBConnectionOptions,
+) => {
+  const client = new MongoClient(options.connectionString);
+  const db = client.db(options.database);
+  const collection = db.collection<EventStream<Event>>(
+    options.collection ?? MongoDBDefaultCollectionName,
+  );
+  await collection.createIndex({ streamName: 1 }, { unique: true });
+  const eventStore = new EventStoreClass(collection);
+  return eventStore;
+};
+
+function executeProjections<EventType extends Event>(
+  params: StreamToProject<EventType>,
+  projections?: Array<Projection<EventType>>,
+) {
+  return Promise.all((projections ?? []).map((project) => project(params)));
+}
+
+function maxEventIndex(
+  expectedStreamVersion?: ExpectedStreamVersion<number>,
+): number | undefined {
+  if (!expectedStreamVersion) return undefined;
+
+  if (typeof expectedStreamVersion === 'number') {
+    return expectedStreamVersion;
+  }
+
+  switch (expectedStreamVersion) {
+    case STREAM_DOES_NOT_EXIST:
+      return 0;
+    default:
+      return undefined;
+  }
+}
+
+/**
+ * Accepts a `streamType` (the type/category of the event stream) and an `entityId`
+ * (the individual entity/object or aggregate ID) and combines them to a singular
+ * `streamName` which can be used in `EventStore`.
+ */
+export function toStreamName<T extends StreamType>(
+  streamType: T,
+  entityId: string,
```

Comment:
**SUGGESTION:** A bit nitpicking, but I'd prefer to call it `streamId`
```suggestion
  streamId: string,
```

Reply:
Note for me: Something for the future is to align this parsing throughout event stores (not this PR).